Add status to dask job

bstadlbauer opened the pull request:
This PR aims at adding some `status` information to the `DaskJob` CRD.

In particular it adds the following fields to `status`: 
- `jobStatus`: Status of the job. string, required. Can by any of: 
    - `JobCreated`: Job CR was created
    - `ClusterCreated`: The `DaskCluster` CR associated with the job has been created
    - `Running`: The job-runner pod is in "Running" state
    - `Successful`: The job-runner pod is in `Succeeded` state
    - `Failed`: The job-runner pod is in the `Failed` statue
- `clusterName`: Name of the `DaskCluster` resorce to be created (string)
- `jobRunnerPodName`: The name of the job-runner pod
- `startTime`: Time when the job-runner pod started running
- `endTime`: Time when the job-runner pod went into the `Succeeded` state

@jacobtomlinson: Happy to change any of the stated/the behavior in general. 

The `test_job` test is also becoming quite large, let me know if I should pull some functions out or split it into multiple tests.

cc @hamersaw

**Edit 2022-10-21**: 
- Split the `Completed` state in into a `Successful` and `Failed` state
- Added `jobRunnerPodName`

Closes #573

jacobtomlinson reviewed (commented):
This looks really great I'm really happy with this functionality.

My only comment is that we are generally preferring `Literal` over `Enum` in other places in Dask, so we might want to do the same here.

Context https://github.com/dask/distributed/pull/6377#discussion_r926953704

bstadlbauer commented:
@jacobtomlinson Thank you for the review!

I've made some small changes: 
- Added a `Failed` state, in the case the `job-runner` pod fails (this was not being accounted for before, so the cluster would also not clean up). I've split the `Completed` state into a `Successful` and `Failed` state respectively. As part of this, I've also slightly refactored the job-test(s) to reduce code duplication.
- Added `jobRunnerPodName` to the status. This could come in hand when getting logs, etc. A few other frameworks do this aswell, so I thought it wouldn't hurt.
- Looked into converting the `enum`'s into  `Literal`s. I have never used `Literal`, but from my limited understanding, this would only help in case there is typing? 🤔 Converted things back to plain strings instead, but happy to change things up

jacobtomlinson reviewed (commented):
Thanks for fixing things up here, those changes sound great.

We should probably do better with type annotations in the controller here, but we can save that for a future PR.

The test failures seem to be related, I think one of the strings may be incorrect.

bstadlbauer commented:
@jacobtomlinson  Just had a look at the tests, I think what's happening is a racecondition there. The cluster has already gone from the `JobCreated` into `ClusterCreated` state before the status was checked - would you have an idea of how to better test this? 🤔 

Also, happy to follow up with type annotations for the controller if that's something that's on the roadmap

jacobtomlinson commented:
> would you have an idea of how to better test this?

I think the main thing being tested is that the states are progressing forwards. If the controller didn't pick it up correctly it would just not have a state. So maybe update the test to assert that state is set to anything.

> Also, happy to follow up with type annotations for the controller if that's something that's on the roadmap

Any code hygiene contributions like this would be appreciated.

bstadlbauer commented:
@jacobtomlinson Good point! Changed the test logic to only assert if `jobStatus` is present. For all other attributes the checks should be fine, as there will only be properties added and their value never changes. 

> Any code hygiene contributions like this would be appreciated.
Good to know! I'll try to get the Flyte plugin across the line, but would be open to some contribution. Will let you know before I would start

jacobtomlinson reviewed (approved):
Awesome many thanks for the efforts here!
